lib/room.cpp

```diff
@@ -2368,14 +2368,39 @@ Room::Changes Room::processStateEvent(const RoomEvent& e)
                 oldStateEvent
                     ? static_cast<const RoomAliasesEvent*>(oldStateEvent)->aliases()
                     : QStringList();
-            connection()->updateRoomAliases(id(), ae.stateKey(),
-                                            previousAliases, ae.aliases());
+            connection()->updateRoomAliases(id(), previousAliases, ae.aliases());
             return OtherChange;
             // clang-format off
         }
-        , [this] (const RoomCanonicalAliasEvent& evt) {
-            setObjectName(evt.alias().isEmpty() ? d->id : evt.alias());
+        , [this, oldStateEvent] (const RoomCanonicalAliasEvent& cae) {
+            // clang-format on
+            QString previousCanonicalAlias =
+                oldStateEvent
+                    ? static_cast<const RoomCanonicalAliasEvent*>(oldStateEvent)
+                          ->alias()
+                    : QString();
+
+            auto previousAltAliases =
+                oldStateEvent
+                    ? static_cast<const RoomCanonicalAliasEvent*>(oldStateEvent)
+                          ->alt_aliases()
+                    : QStringList();
+
+            if (!previousCanonicalAlias.isEmpty()) {
+                previousAltAliases.push_back(previousCanonicalAlias);
+            }
+
+            const auto previousAliases = std::move(previousAltAliases);
+
+            auto newAliases = cae.alt_aliases();
+
+            if (!cae.alias().isEmpty()) {
+                newAliases.push_front(cae.alias());
+            }
+
+            connection()->updateRoomAliases(id(), previousAliases, newAliases);
```

Comment:
You seem to have forgotten to incorporate the original code reacting to the event (the `setObjectName` line).